nixos/modules/virtualisation/ec2-metadata-fetcher.nix

```diff
@@ -1,23 +1,58 @@
-{ targetRoot, wgetExtraOptions }:
+{ curl, targetRoot, wgetExtraOptions }:
+# Note: be very cautious about dependencies, each dependency grows
+# the closure of the initrd. Ideally we would not even require curl,
+# but there is no reasonable way to send an HTTP PUT request without
+# it. Note: do not be fooled: the wget referenced in this script
+# is busybox's wget, not the fully featured one with --method support.
+#
+# Make sure that every package you depend on here is already listed as
+# a channel blocker for both the full-sized and small channels.
+# Otherwise, we risk breaking user deploys in released channels.
 ''
   metaDir=${targetRoot}etc/ec2-metadata
   mkdir -m 0755 -p "$metaDir"
 
+  get_imds_token() {
+    # retry-delay of 1 selected to give the system a second to get going,
+    # but not add a lot to the bootup time
+    ${curl}/bin/curl \
+      -v \
+      --retry 3 \
+      --retry-delay 1 \
```

Comment:
Yeah. It could easily succeed within a second or so. Blocking early boot for 30s at a time seems intense.

Reply:
Is there a chance to *not* have v2 enabled where this will always fail on some AWS account? If that is the case we might want to check what the error values there are.

If the metadata service wasn't available just yet would it be worse to delay the boot (and be useless?) or to delay it by 30s and have it potentially recover?

Reply:
> Is there a chance to not have v2 enabled where this will always fail on some AWS account? If that is the case we might want to check what the error values there are.

No. You can only have V1 & V2 _or_ V2 only: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html#instance-metadata-v2-how-it-works

> If the metadata service wasn't available just yet would it be worse to delay the boot (and be useless?) or to delay it by 30s and have it potentially recover?

The instance metadata service is never not available, and is provided by the hypervisor. The race here is "did we setup the network yet?" In general, I think this almost-never fails past the initial request, and so giving it up to 9 seconds seems generous already.

Reply:
Instead of keeping the existing connect-timeout of 30s (woof) I'd rather extend the retry loop to be like 30 tries or something. But also I feel comfortable that the current configuration is good (or I wouldn't have PR'd it :sweat_smile: )